GooglePlayInstant/Editor/AndroidManifestHelper.cs

```diff
@@ -0,0 +1,241 @@
+                new XElement(Application,
+                    new XElement(Activity,
+                        new XAttribute(AndroidNameXName, "com.unity3d.player.UnityPlayerActivity"),
+                        new XElement(IntentFilter,
+                            new XElement(Action, new XAttribute(AndroidNameXName, ActionMain)),
+                            new XElement(Category, new XAttribute(AndroidNameXName, CategoryLauncher)))))));
+        }
+
+        public static void ConvertManifestToInstalled(XDocument doc)
+        {
+            foreach (var manifestElement in doc.Elements(Manifest))
+            {
+                manifestElement.Attributes(AndroidTargetSandboxVersionXName).Remove();
+                foreach (var applicationElement in manifestElement.Elements(Application))
+                {
+                    foreach (var mainActivity in GetMainActivities(applicationElement))
+                    {
+                        // TODO: also remove view intent filters?
+                        GetDefaultUrlMetaDataElements(mainActivity).Remove();
+                    }
+                }
+            }
+        }
+
+        public static string ConvertManifestToInstant(XDocument doc, Uri uri)
+        {
+            var manifestElement = GetExactlyOne(doc.Elements(Manifest));
+            if (manifestElement == null)
+            {
+                return PreconditionOneManifestElement;
+            }
+
+            var androidAttribute = manifestElement.Attribute(AndroidXmlns);
+            if (androidAttribute == null)
+            {
+                return PreconditionMissingXmlnsAndroid;
+            }
+
+            if (androidAttribute.Value != AndroidNamespaceUrl)
+            {
+                return PreconditionInvalidXmlnsAndroid;
+            }
+
+            // TSV2 is required for instant apps starting with Android Oreo.
+            manifestElement.SetAttributeValue(AndroidTargetSandboxVersionXName, "2");
+
+            return uri == null ? null : AddDefaultUrl(manifestElement, uri);
+        }
+
+        private static string AddDefaultUrl(XElement manifestElement, Uri uri)
+        {
+            var applicationElement = GetExactlyOne(manifestElement.Elements(Application));
+            if (applicationElement == null)
+            {
+                return PreconditionOneApplicationElement;
+            }
+
+            var mainActivity = GetExactlyOne(GetMainActivities(applicationElement));
+            if (mainActivity == null)
+            {
+                return PreconditionOneMainActivity;
+            }
+
+            var updateViewIntentFilterResult = UpdateViewIntentFilter(mainActivity, uri);
+            if (updateViewIntentFilterResult != null)
+            {
+                return updateViewIntentFilterResult;
+            }
+
+            return UpdateDefaultUrlElement(mainActivity, uri);
+        }
+
+        private static string UpdateViewIntentFilter(XElement mainActivity, Uri uri)
```

Comment:
Can you add docs for non-trivial methods.

Reply:
Added /// style XML comments to most public classes and methods.

Added // style comments to anything that didn't seem patently clear.